:sparkles: feat: Support multi-types parameters (#3)
LeadSimple/boring-service: README.md

Reviewer: Last thing, I promise. 😄 

Could you update the README to show an example of both a single type and an array of types? And actually, I should have had a proc example in there from the beginning as well (my bad). And now that I'm looking, we should probably have an example of no type checking as well (again, my bad).

What about something like:

```ruby
class RandomService < BoringService
  # Type checking is optional. If omitted, anything is accepted.
  parameter :random_object

  def call
    random_object&.class
  end
end

# The class-method version of `call` accepts the arguments as named parameters and, subsequently, calls the
# instance-method version of `call` (as it's defined in your service class).
RandomService.call                        #=> raise BoringService::ParameterRequired
RandomService.call(random_object: 'test') #=> String
RandomService.call(random_object: 1)      #=> Integer

class CalculationService < BoringService
  # If a type checking Class is provided, the type check will be enforced against the defined parameter.
  parameter :start_number, Integer

  # Parameters that define a `default` are optional.
  parameter :end_number, Integer, default: 2

  def call
    @magic_number = 42
    perform_complex_calculation
  end

  private

  def perform_complex_calculation
    # The arguments are available as accessors
    start_number + end_number + @magic_number
  end
end

CalculationService.call(start_number: 1, end_number: 3) #=> 46
CalculationService.call(start_number: 1)                #=> 45
CalculationService.call(start_number: '1')              #=> BoringService::InvalidParameterValue
CalculationService.call(end_number: 3)                  #=> raise BoringService::ParameterRequired

class MoreOptionsService < BoringService
  # Type checking can also be done with an Array of classes, a proc/lambda, or anything that responds to #===.
  parameter :times_to_run, [String, Integer]
  parameter :return_output, -> (param) { param.in?([true, false]) }

  # `default` also supports a proc/lambda, that gets evaluated at object instantiation.
  parameter :timestamp, Time, default: -> { Time.now.beginning_of_day }

  def call
    Array.new(times_to_run.to_i) { timestamp if return_output }
  end
end

MoreOptionsService.call(times_to_run: 1, return_output: true)    #=> [2024-08-30 00:00:00 -0400]
MoreOptionsService.call(times_to_run: 1, return_output: false)   #=> [nil]
MoreOptionsService.call(times_to_run: 5, return_output: false)   #=> [nil, nil, nil, nil, nil]
MoreOptionsService.call(times_to_run: '5', return_output: false) #=> [nil, nil, nil, nil, nil]
MoreOptionsService.call(times_to_run: 5, return_output: 'false') #=> raise BoringService::InvalidParameterValue
MoreOptionsService.call(
  times_to_run: 1, 
  return_output: true, 
  timestamp: Time.now.beginning_of_day - 1.day
)                                                                #=> [2024-08-29 00:00:00 -0400]
```

Author: funny enough, I realized to add a test with the new configuration/variants but didn't consider adding it to the readme too 😅 

gonna be committing right away

Author: > Last thing, I promise. 😄

no please, don't ever hesitate to leave more feedback in PR reviews: it's always welcome 😄 
btw I just addressed the Markdown changes with better examples and explanations @jeffdill2

```ruby
class CalculationService < BoringService
  # Type checking is optional. If omitted, anything is accepted.

>   # Type checking can also be done with a proc or anything that responds to #=== (and even an array of it!)

Reviewer: ```suggestion
  # Type checking can also be done with an Array of classes, a proc, or anything that responds to #===
```

Author: oh perfect, this reads better 😄

  # e.g. parameter :start_number, ->(p) { p.respond_to?(:to_i) }

>   parameter :start_number, [Integer, Float], required: true

Reviewer: ```suggestion
  parameter :start_number, Integer
  parameter :random_number_or_string, [Integer, String]
```

Reviewer: And then the rest of the example below would need to reflect the addition of the param that can be either a number or a string.

Reviewer: And I removed the `required: true`, since that's not a param.

Author: > And I removed the required: true, since that's not a param.

actually, you're showing an example below where you raise a `BoringService::ParameterRequired` error when calling the `CalculationService` without `start_number`. I was actually fixing your test class in the doc to properly reflex the documentation examples 🤔 

<img width="818" alt="image" src="https://github.com/user-attachments/assets/a2242f28-511f-4d69-9922-f09d7717dbd6">

or am I missing something @jeffdill2 ?

Author: oh, my bad my bad
you're actually considering every parameter as required unless it has a default value 💡 

we may want to support a `required` options - but that's an eventual future issue/PR, not for today 😄

Reviewer: It functions the same way as vanilla Ruby (intentionally). I.e. if you define a parameter, you must provide it, unless you define a default value, in which case you don't have to provide it.
